h2o-automl/src/main/java/ai/h2o/automl/preprocessing/TargetEncoding.java

```diff
@@ -0,0 +1,126 @@
+package ai.h2o.automl.preprocessing;
+
+import ai.h2o.automl.AutoML;
+import ai.h2o.targetencoding.TargetEncoder;
+import ai.h2o.targetencoding.TargetEncoderModel;
+import ai.h2o.targetencoding.TargetEncoderModel.DataLeakageHandlingStrategy;
+import ai.h2o.targetencoding.TargetEncoderModel.TargetEncoderParameters;
+import ai.h2o.targetencoding.TargetEncoderPreprocessor;
+import hex.Model;
+import hex.Model.Parameters.FoldAssignmentScheme;
+import hex.ModelPreprocessor;
+import water.DKV;
+import water.Key;
+import water.fvec.Frame;
+import water.fvec.Vec;
+import water.rapids.ast.prims.advmath.AstKFold;
+import water.util.ArrayUtils;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class TargetEncoding implements PreprocessingStep {
+    
+    private transient AutoML _aml;
+    private transient TargetEncoderPreprocessor _tePreprocessor;
+    private transient TargetEncoderModel _teModel;
+    private final transient List<Completer> _disposables = new ArrayList<>();
+
+    public TargetEncoding(AutoML aml) {
+        _aml = aml;
+    }
+
+    @Override
+    public void prepare() {
+        TargetEncoderParameters params = new TargetEncoderParameters();
+        params._train = _aml.getTrainingFrame()._key;
+        params._response_column = _aml.getBuildSpec().input_spec.response_column;
+        params._keep_original_features = false;
+        params._blending = true;
+        params._noise = 0;
+        params._seed = _aml.getBuildSpec().build_control.stopping_criteria.seed();
+        
+        if (_aml.isCVEnabled()) {
+            params._data_leakage_handling = DataLeakageHandlingStrategy.KFold;
+            params._fold_column = _aml.getBuildSpec().input_spec.fold_column;
+            if (params._fold_column == null) {
+                //generate fold column
+                Frame train = new Frame(params.train());
+                Vec foldColumn = createFoldColumn(
+                        params.train(), 
+                        FoldAssignmentScheme.Modulo,
```

Comment:
using Modulo because that's we use for CV in AutoML: hardcoded, @erin wasn't it something we wanted to change?

Reply:
@sebhrusen We put this here originally (I think) because we wanted all the models to have the exact same folds (even if seed was not set).  However, I think the Stacked Ensembles should be the same (or slightly more robust even), if we use different CV folds because there may be higher variance across the base models.

I guess the real question is -- do we need the CV folds used in TE to be the same folds used in the base models?  If so, then we should keep modulo; if not, then we can remove it.

Reply:
> do we need the CV folds used in TE to be the same folds used in the base models?

With current integration of TE, this is necessary to use the same folds: this allows to build the CV models without  introducing data leakage from the target encoded features...  quickly, it allows to behave as if for each CV model, we were training a new TE model on the cv-train data, and apply it separately then to both cv-train and cv-test for out-of-fold predictions.

Reply:
Hopefully I won't be off topic here.
Best practice for Target Encoding is nested CV to avoid any sort of leakage and that would like this in plain python : 

```
# NOTE : categorical_list contains the list of categorical features
folds = KFold()
nested_folds = KFold()

for trn_, val_ in folds.split(train_frame):
    # Standard fold split
    trn_x, trn_y = train_frame[trn_], y[trn_]
    val_x,  val_y = train_frame[val_], y[val_]

    # Nested CV 
    cat_te_trn = np.zeros((trn_x.shape[0], len(cat_features))
    cat_te_val = np.zeros((val_x.shape[0], len(cat_features))
    for nested_trn_, nested_val_ in nested_folds.split(trn_x):
        nested_trn_x, nested_trn_y = trn_x[nested_trn_], trn_y[nested_trn_]
        nested_val_x, nested_val_y = trn_x[nested_val_], trn_y[nested_val_]
        
        for i_c, cat in enumerate(cat_features):
            model = TEModel().fit(nested_trn_x[cat], nested_trn_y)
            cat_te_trn[nested_val_, i_c] = model.transform(nested_val_x[cat])
            cat_te_val[val_, i_c] += model.transform(val_x[cat]) / folds.n_splits

     # Now you can use cat_te_tran and cat_te_val to train the model
     ...
```
Hopefully I did not get the code wrong.

Reply:
To create `cat_te_val` you could fit a TEModel on `trn_x` directly instead of averaging btw.

Reply:
@sebhrusen Regarding the modulo folds (so that we can ensure same folds used in CV and modeling) – another option to handle this is to add an error if the user does not set the `seed` arg when TE is turned on.  If the user does set the seed, then we can ensure same folds across TE and modeling, right?  

Maybe i am not thinking about this correctly, but if a user explicitly sets the seed, and also turns on TE... they will get different folds than if they had just set seed and not turned on TE, right?  That will make runs difficult to compare (TE on/off), so that's why I am wondering if it's better to force the user to set a seed when using TE (or providing a `fold_column`)?  This would force the user to be reproducible when TE is turned on. 

Thoughts?

Reply:
@sebhrusen I created a ticket here to revisit whether or not we need modulo folds (vs forcing a seed to be set): https://0xdata.atlassian.net/browse/PUBDEV-7801